File: 04-trees-and-graphs/01_binary_tree_traversals/binary_tree_traversals.py
"""
Problem: Binary Tree Traversals (Inorder, Preorder, Postorder, Level Order)

Time Complexity: O(N)
Space Complexity: O(H) recursion, O(W) queue
"""

from collections import deque
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inorder(root: Optional[TreeNode]) -> List[int]:
    return inorder(root.left) + [root.val] + inorder(root.right) if root else []

def preorder(root: Optional[TreeNode]) -> List[int]:
    return [root.val] + preorder(root.left) + preorder(root.right) if root else []

def postorder(root: Optional[TreeNode]) -> List[int]:
    return postorder(root.left) + postorder(root.right) + [root.val] if root else []

def level_order(root: Optional[TreeNode]) -> List[int]:
    if not root:
        return []
    res, q = [], deque([root])
    while q:
        curr = q.popleft()
        res.append(curr.val)
        if curr.left:
            q.append(curr.left)
        if curr.right:
            q.append(curr.right)
    return res

if __name__ == "__main__":
    print("[Python] Binary Tree Traversals Test")
    root = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))
    print("Preorder:   ", preorder(root))
    print("Inorder:    ", inorder(root))
    print("Postorder:  ", postorder(root))
    print("Level-Order:", level_order(root))
    assert inorder(root) == [4, 2, 5, 1, 3]
    print("Tree traversals test passed!")

File: 04-trees-and-graphs/01_binary_tree_traversals/binary_tree_traversals.cpp
/**
 * Problem: Binary Tree Traversals
 * 
 * Time Complexity: O(N)
 * Space Complexity: O(H) recursion stack, O(W) BFS queue
 */

#include <iostream>
#include <queue>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

void inorder(TreeNode* root) {
    if (!root) return;
    inorder(root->left);
    std::cout << root->val << " ";
    inorder(root->right);
}

void preorder(TreeNode* root) {
    if (!root) return;
    std::cout << root->val << " ";
    preorder(root->left);
    preorder(root->right);
}

void postorder(TreeNode* root) {
    if (!root) return;
    postorder(root->left);
    postorder(root->right);
    std::cout << root->val << " ";
}

void levelOrder(TreeNode* root) {
    if (!root) return;
    std::queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* curr = q.front();
        q.pop();
        std::cout << curr->val << " ";
        if (curr->left) q.push(curr->left);
        if (curr->right) q.push(curr->right);
    }
}

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    std::cout << "[C++] Binary Tree Traversals\n";
    std::cout << "Preorder:   "; preorder(root); std::cout << "\n";
    std::cout << "Inorder:    "; inorder(root); std::cout << "\n";
    std::cout << "Postorder:  "; postorder(root); std::cout << "\n";
    std::cout << "Level-Order:"; levelOrder(root); std::cout << "\n";
    return 0;
}
